File: DailyCore/Sources/DailyCore/Services/TagdosParser.swift
import Foundation

/// Fast, deterministic parser for the TagDoS syntax.
/// Syntactic rules:
/// - Streams are collections of clusters separated by ` & `
/// - Clusters are sequences of mental action tags separated by `/`
/// - Semantic tokens: `$` or `€` (financial), `!` (urgent), numbers (metric/date)
public struct TagdosParser: Sendable {
    public static let shared = TagdosParser()

    public init() {}

    /// Parses a raw TagDoS line into a fully populated `TagDoStream`.
    public func parseStream(
        rawText: String,
        title: String,
        id: UUID = UUID(),
        streamReminder: Date? = nil,
        orderIndex: Int = 0,
        existingStream: TagDoStream? = nil
    ) -> TagDoStream {
        let trimmed = rawText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            return TagDoStream(
                id: id,
                title: title,
                rawText: "",
                clusters: [],
                streamReminder: streamReminder,
                orderIndex: orderIndex
            )
        }

        // Split by cluster separator: ` & ` (support variable spacing around `&`)
        let rawClusters = trimmed.components(separatedBy: "&")
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }

        var parsedClusters: [TagDoCluster] = []

        for rawCluster in rawClusters {
            // Split cluster by pill separator: `/`
            let rawPillTokens = rawCluster.components(separatedBy: "/")
                .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
                .filter { !$0.isEmpty }

            var pills: [TagDoPill] = []
            for token in rawPillTokens {
                // If an existing pill with the same token exists, preserve its completion and notes state
                if let existing = existingStream?.allPills.first(where: { $0.rawText == token }) {
                    pills.append(TagDoPill(
                        id: existing.id,
                        rawText: token,
                        type: existing.type,
                        isCompleted: existing.isCompleted,
                        customReminderDate: existing.customReminderDate,
                        decodedNote: existing.decodedNote
                    ))
                } else {
                    pills.append(TagDoPill(rawText: token))
                }
            }

            if !pills.isEmpty {
                parsedClusters.append(TagDoCluster(pills: pills, rawText: rawCluster))
            }
        }

        return TagDoStream(
            id: id,
            title: title,
            rawText: trimmed,
            clusters: parsedClusters,
            streamReminder: streamReminder,
            orderIndex: orderIndex
        )
    }

    /// Serializes a `TagDoStream` back to its canonical raw string format.
    public func serializeStream(_ stream: TagDoStream) -> String {
        let clusterStrings = stream.clusters.map { cluster in
            cluster.pills.map { $0.rawText }.joined(separator: "/")
        }
        return clusterStrings.joined(separator: " & ")
    }

    // MARK: - Interactive Mutations

    /// Cycles a pill to the end of its cluster (or end of stream) for recurring habits.
    public func recyclePillToBack(in stream: TagDoStream, pillId: UUID) -> TagDoStream {
        var updatedClusters = stream.clusters

        for clusterIndex in 0..<updatedClusters.count {
            var cluster = updatedClusters[clusterIndex]
            if let pillIndex = cluster.pills.firstIndex(where: { $0.id == pillId }) {
                var pill = cluster.pills.remove(at: pillIndex)
                pill.isCompleted = false // Reset completed status for next cycle
                cluster.pills.append(pill)
                cluster.rawText = cluster.pills.map { $0.rawText }.joined(separator: "/")
                updatedClusters[clusterIndex] = cluster
                break
            }
        }

        let newRawText = updatedClusters.map { $0.rawText }.joined(separator: " & ")
        return TagDoStream(
            id: stream.id,
            title: stream.title,
            rawText: newRawText,
            clusters: updatedClusters,
            streamReminder: stream.streamReminder,
            orderIndex: stream.orderIndex
        )
    }

    /// Toggles completion status of a pill in-place.
    public func togglePillCompletion(in stream: TagDoStream, pillId: UUID) -> TagDoStream {
        var updatedClusters = stream.clusters

        for clusterIndex in 0..<updatedClusters.count {
            var cluster = updatedClusters[clusterIndex]
            if let pillIndex = cluster.pills.firstIndex(where: { $0.id == pillId }) {
                cluster.pills[pillIndex].isCompleted.toggle()
                updatedClusters[clusterIndex] = cluster
                break
            }
        }

        return TagDoStream(
            id: stream.id,
            title: stream.title,
            rawText: stream.rawText,
            clusters: updatedClusters,
            streamReminder: stream.streamReminder,
            orderIndex: stream.orderIndex
        )
    }

    /// Removes a pill from the stream (one-off task).
    public func removePill(in stream: TagDoStream, pillId: UUID) -> TagDoStream {
        var updatedClusters = stream.clusters

        for clusterIndex in (0..<updatedClusters.count).reversed() {
            var cluster = updatedClusters[clusterIndex]
            if let pillIndex = cluster.pills.firstIndex(where: { $0.id == pillId }) {
                cluster.pills.remove(at: pillIndex)
                if cluster.pills.isEmpty {
                    updatedClusters.remove(at: clusterIndex)
                } else {
                    cluster.rawText = cluster.pills.map { $0.rawText }.joined(separator: "/")
                    updatedClusters[clusterIndex] = cluster
                }
                break
            }
        }

        let newRawText = updatedClusters.map { $0.rawText }.joined(separator: " & ")
        return TagDoStream(
            id: stream.id,
            title: stream.title,
            rawText: newRawText,
            clusters: updatedClusters,
            streamReminder: stream.streamReminder,
            orderIndex: stream.orderIndex
        )
    }

    /// Moves a pill to the very beginning of its cluster (highest priority).
    public func movePillToFront(in stream: TagDoStream, pillId: UUID) -> TagDoStream {
        var updatedClusters = stream.clusters

        for clusterIndex in 0..<updatedClusters.count {
            var cluster = updatedClusters[clusterIndex]
            if let pillIndex = cluster.pills.firstIndex(where: { $0.id == pillId }) {
                let pill = cluster.pills.remove(at: pillIndex)
                cluster.pills.insert(pill, at: 0)
                cluster.rawText = cluster.pills.map { $0.rawText }.joined(separator: "/")
                updatedClusters[clusterIndex] = cluster
                break
            }
        }

        let newRawText = updatedClusters.map { $0.rawText }.joined(separator: " & ")
        return TagDoStream(
            id: stream.id,
            title: stream.title,
            rawText: newRawText,
            clusters: updatedClusters,
            streamReminder: stream.streamReminder,
            orderIndex: stream.orderIndex,
            activeMemos: stream.activeMemos,
            attachments: stream.attachments,
            updatedAt: Date()
        )
    }

    /// Adds a new pill into a specific cluster or creates a new one.
    public func addPill(in stream: TagDoStream, clusterIndex: Int, text: String) -> TagDoStream {
        let cleanText = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !cleanText.isEmpty else { return stream }

        var updatedClusters = stream.clusters
        let newPill = TagDoPill(rawText: cleanText)

        if updatedClusters.isEmpty {
            updatedClusters.append(TagDoCluster(pills: [newPill], rawText: cleanText))
        } else if clusterIndex >= 0 && clusterIndex < updatedClusters.count {
            var cluster = updatedClusters[clusterIndex]
            cluster.pills.append(newPill)
            cluster.rawText = cluster.pills.map { $0.rawText }.joined(separator: "/")
            updatedClusters[clusterIndex] = cluster
        } else {
            if var last = updatedClusters.last {
                last.pills.append(newPill)
                last.rawText = last.pills.map { $0.rawText }.joined(separator: "/")
                updatedClusters[updatedClusters.count - 1] = last
            } else {
                updatedClusters.append(TagDoCluster(pills: [newPill], rawText: cleanText))
            }
        }

        let newRawText = updatedClusters.map { $0.rawText }.joined(separator: " & ")
        return TagDoStream(
            id: stream.id,
            title: stream.title,
            rawText: newRawText,
            clusters: updatedClusters,
            streamReminder: stream.streamReminder,
            orderIndex: stream.orderIndex,
            activeMemos: stream.activeMemos,
            attachments: stream.attachments,
            updatedAt: Date()
        )
    }
}

File: Android/core-model/src/main/java/com/intellidream/daily/model/TagdosParser.kt
package com.intellidream.daily.model

import java.util.UUID

/**
 * Fast, deterministic parser for the TagDoS syntax.
 * 1:1 Kotlin port of iOS DailyCore [TagdosParser.swift].
 *
 * Syntactic rules:
 * - Streams are collections of clusters separated by ` & `
 * - Clusters are sequences of mental action tags separated by `/`
 * - Semantic tokens: `$` or `€` (financial), `!` (urgent), digits (metric/date)
 */
object TagdosParser {

    /**
     * Parses a raw TagDoS line into a fully populated [TagDoStream].
     */
    fun parseStream(
        rawText: String,
        title: String,
        id: String = UUID.randomUUID().toString(),
        streamReminder: Long? = null,
        orderIndex: Int = 0,
        existingStream: TagDoStream? = null
    ): TagDoStream {
        val trimmed = rawText.trim()
        if (trimmed.isEmpty()) {
            return TagDoStream(
                id = id,
                title = title,
                customTitle = existingStream?.customTitle,
                rawText = "",
                clusters = emptyList(),
                streamReminder = streamReminder,
                orderIndex = orderIndex,
                activeMemos = existingStream?.activeMemos ?: "",
                attachments = existingStream?.attachments ?: emptyList()
            )
        }

        // Split by cluster separator: ` & ` (support variable spacing around `&`)
        val rawClusters = trimmed.split("&")
            .map { it.trim() }
            .filter { it.isNotEmpty() }

        val parsedClusters = mutableListOf<TagDoCluster>()

        for (rawCluster in rawClusters) {
            // Split cluster by pill separator: `/`
            val rawPillTokens = rawCluster.split("/")
                .map { it.trim() }
                .filter { it.isNotEmpty() }

            val pills = mutableListOf<TagDoPill>()
            for (token in rawPillTokens) {
                // If an existing pill with the same token exists, preserve its completion and notes state
                val existing = existingStream?.allPills?.firstOrNull { it.rawText == token }
                if (existing != null) {
                    pills.add(
                        TagDoPill(
                            id = existing.id,
                            rawText = token,
                            type = existing.type,
                            isCompleted = existing.isCompleted,
                            customReminderDate = existing.customReminderDate,
                            decodedNote = existing.decodedNote
                        )
                    )
                } else {
                    pills.add(TagDoPill(rawText = token))
                }
            }

            if (pills.isNotEmpty()) {
                parsedClusters.add(TagDoCluster(pills = pills, rawText = rawCluster))
            }
        }

        return TagDoStream(
            id = id,
            title = title,
            customTitle = existingStream?.customTitle,
            rawText = trimmed,
            clusters = parsedClusters,
            streamReminder = streamReminder,
            orderIndex = orderIndex,
            activeMemos = existingStream?.activeMemos ?: "",
            attachments = existingStream?.attachments ?: emptyList()
        )
    }

    /**
     * Serializes a [TagDoStream] back to its canonical raw string format.
     */
    fun serializeStream(stream: TagDoStream): String {
        return stream.clusters.joinToString(" & ") { cluster ->
            cluster.pills.joinToString("/") { it.rawText }
        }
    }

    // MARK: - Interactive Mutations

    /**
     * Cycles a pill to the end of its cluster for recurring habits/tasks.
     * Resets completed status for the next cycle.
     */
    fun recyclePillToBack(stream: TagDoStream, pillId: String): TagDoStream {
        val updatedClusters = stream.clusters.map { it.copy(pills = it.pills.toMutableList()) }.toMutableList()

        for (clusterIndex in updatedClusters.indices) {
            val cluster = updatedClusters[clusterIndex]
            val pills = cluster.pills.toMutableList()
            val pillIndex = pills.indexOfFirst { it.id == pillId }
            if (pillIndex != -1) {
                val pill = pills.removeAt(pillIndex)
                pills.add(pill.copy(isCompleted = false))
                val newRawText = pills.joinToString("/") { it.rawText }
                updatedClusters[clusterIndex] = cluster.copy(pills = pills, rawText = newRawText)
                break
            }
        }

        val newRawText = updatedClusters.joinToString(" & ") { it.rawText }
        return stream.copy(
            rawText = newRawText,
            clusters = updatedClusters,
            updatedAt = System.currentTimeMillis()
        )
    }

    /**
     * Toggles completion status of a pill in-place.
     */
    fun togglePillCompletion(stream: TagDoStream, pillId: String): TagDoStream {
        val updatedClusters = stream.clusters.map { it.copy(pills = it.pills.toMutableList()) }.toMutableList()

        for (clusterIndex in updatedClusters.indices) {
            val cluster = updatedClusters[clusterIndex]
            val pills = cluster.pills.toMutableList()
            val pillIndex = pills.indexOfFirst { it.id == pillId }
            if (pillIndex != -1) {
                val pill = pills[pillIndex]
                pills[pillIndex] = pill.copy(isCompleted = !pill.isCompleted)
                updatedClusters[clusterIndex] = cluster.copy(pills = pills)
                break
            }
        }

        return stream.copy(
            clusters = updatedClusters,
            updatedAt = System.currentTimeMillis()
        )
    }

    /**
     * Removes a pill from the stream (one-off task).
     */
    fun removePill(stream: TagDoStream, pillId: String): TagDoStream {
        val updatedClusters = stream.clusters.map { it.copy(pills = it.pills.toMutableList()) }.toMutableList()

        for (clusterIndex in updatedClusters.indices.reversed()) {
            val cluster = updatedClusters[clusterIndex]
            val pills = cluster.pills.toMutableList()
            val pillIndex = pills.indexOfFirst { it.id == pillId }
            if (pillIndex != -1) {
                pills.removeAt(pillIndex)
                if (pills.isEmpty()) {
                    updatedClusters.removeAt(clusterIndex)
                } else {
                    val newRawText = pills.joinToString("/") { it.rawText }
                    updatedClusters[clusterIndex] = cluster.copy(pills = pills, rawText = newRawText)
                }
                break
            }
        }

        val newRawText = updatedClusters.joinToString(" & ") { it.rawText }
        return stream.copy(
            rawText = newRawText,
            clusters = updatedClusters,
            updatedAt = System.currentTimeMillis()
        )
    }

    /**
     * Moves a pill to the very beginning of its cluster (highest priority).
     */
    fun movePillToFront(stream: TagDoStream, pillId: String): TagDoStream {
        val updatedClusters = stream.clusters.map { it.copy(pills = it.pills.toMutableList()) }.toMutableList()

        for (clusterIndex in updatedClusters.indices) {
            val cluster = updatedClusters[clusterIndex]
            val pills = cluster.pills.toMutableList()
            val pillIndex = pills.indexOfFirst { it.id == pillId }
            if (pillIndex != -1) {
                val pill = pills.removeAt(pillIndex)
                pills.add(0, pill)
                val newRawText = pills.joinToString("/") { it.rawText }
                updatedClusters[clusterIndex] = cluster.copy(pills = pills, rawText = newRawText)
                break
            }
        }

        val newRawText = updatedClusters.joinToString(" & ") { it.rawText }
        return stream.copy(
            rawText = newRawText,
            clusters = updatedClusters,
            updatedAt = System.currentTimeMillis()
        )
    }

    /**
     * Adds a new pill into a specific cluster or creates a new one.
     */
    fun addPill(stream: TagDoStream, clusterIndex: Int, text: String): TagDoStream {
        val cleanText = text.trim()
        if (cleanText.isEmpty()) return stream

        val updatedClusters = stream.clusters.map { it.copy(pills = it.pills.toMutableList()) }.toMutableList()
        val newPill = TagDoPill(rawText = cleanText)

        if (updatedClusters.isEmpty()) {
            updatedClusters.add(TagDoCluster(pills = listOf(newPill), rawText = cleanText))
        } else if (clusterIndex in 0 until updatedClusters.count()) {
            val cluster = updatedClusters[clusterIndex]
            val pills = cluster.pills.toMutableList()
            pills.add(newPill)
            val newRawText = pills.joinToString("/") { it.rawText }
            updatedClusters[clusterIndex] = cluster.copy(pills = pills, rawText = newRawText)
        } else {
            val last = updatedClusters.last()
            val pills = last.pills.toMutableList()
            pills.add(newPill)
            val newRawText = pills.joinToString("/") { it.rawText }
            updatedClusters[updatedClusters.count() - 1] = last.copy(pills = pills, rawText = newRawText)
        }

        val newRawText = updatedClusters.joinToString(" & ") { it.rawText }
        return stream.copy(
            rawText = newRawText,
            clusters = updatedClusters,
            updatedAt = System.currentTimeMillis()
        )
    }
}
